File: 2501-3000/2561.Rearranging Fruits.ts
/*
 要么用最小的min当工具人，连续换两次，以交换其他两个数；
 要么用array里i指向的最小值换array后半部分较大的。
 basket1和basket2中需要被换的数，可以一股脑放进array里，无需考虑是从哪个basket换到哪个basket。
*/
function minCost(basket1: number[], basket2: number[]): number {
    const N = basket1.length;
    const counter: {[key: number] : number} = {};
    for (let i = 0; i < N; i++) {
        counter[basket1[i]] ||= 0;
        counter[basket1[i]]++;
        counter[basket2[i]] ||= 0;
        counter[basket2[i]]--;
    }
    const min = Math.min(...Object.keys(counter).map(num => parseInt(num)));
    const array: number[] = [];
    if (Object.values(counter).some(value => value % 2)) {
        return -1;
    }
    Object.entries(counter).sort((a, b) => parseInt(a[0]) - parseInt(b[0])).forEach(([key, value]) => {
        const num = parseInt(key);
        for (let i = 0; i < Math.abs(value) / 2; i++) {
            array.push(num);
        }
    });
    const half = array.slice(0, array.length / 2);
    return half.reduce((acc, num) => acc + Math.min(min * 2, num), 0);
};

File: 2501-3000/2561.Rearranging Fruits.py
from collections import deque
from collections import Counter

class Solution:
    def minCost(self, basket1: List[int], basket2: List[int]) -> int:
        cnt = Counter(basket1) + Counter(basket2)
        for value in cnt.values():
            if value % 2 == 1:
                return -1
        n = len(basket1)
        basket1.sort()
        basket2.sort()
        candidates1 = deque()
        candidates2 = deque()
        stay = None
        i = j = 0
        while i < n and j < n:
            if basket1[i] == basket2[j]:
                if stay is None:
                    stay = basket1[i]
                i += 1
                j += 1
            elif basket1[i] < basket2[j]:
                candidates1.append(basket1[i])
                i += 1
            else:
                candidates2.append(basket2[j])
                j += 1
        if i < n:
            candidates1.extend(basket1[i:])
        if j < n:
            candidates2.extend(basket2[j:])
        
            
        if stay is None:
            stay = float("inf")
        res = 0
        while candidates1 and candidates2:
            c1 = candidates1[0]
            c2 = candidates2[0]
            if stay is not None and stay * 2 < min(c1, c2):
                res += stay * len(candidates1)
                break
            elif c1 < c2:
                res += c1
                candidates1.popleft()
                candidates1.popleft()
                candidates2.pop()
                candidates2.pop()
                stay = min(stay, c1)
            else:
                res += c2
                candidates1.pop()
                candidates1.pop()
                candidates2.popleft()
                candidates2.popleft()
                stay = min(stay, c2)
        return res

